src/dos/dos_locale.cpp

```diff
@@ -0,0 +1,2904 @@
+	destination[offset + 1] = 0;
+
+	offset = InfoOffsetDecimalSeparator;
+	destination[offset + 0] = static_cast<uint8_t>(source.decimal_separator);
+	destination[offset + 1] = 0;
+
+	offset = InfoOffsetListSeparator;
+	destination[offset + 0] = static_cast<uint8_t>(source.list_separator);
+	destination[offset + 1] = 0;
+
+	// Set currency symbol and format
+
+	assert(source.currency_code.size() < MaxCurrencySymbolLength);
+	memset(&destination[InfoOffsetCurrencySymbol], 0, MaxCurrencySymbolLength + 1);
+
+	bool found = false;
+	for (const auto& candidate_utf8 : source.currency_symbols_utf8) {
+		std::string candidate = {};
+
+		// Check if the currency can be converted to current code page
+
+		if (!utf8_to_dos(candidate_utf8, candidate, UnicodeFallback::Null)) {
+			continue;
+		}
+		if (candidate.length() > MaxCurrencySymbolLength) {
+			continue;
+		}
+
+		found = true;
+		for (const auto character : candidate) {
+			if (character == 0) {
+				found = false;
+				break;
+			}
+		}
+
+		if (found) {
+			memcpy(&destination[InfoOffsetCurrencySymbol],
+			       candidate.c_str(),
+			       candidate.length());
+			break;
+		}
+	}
+
+	offset = InfoOffsetCurrencyFormat;
+	destination[offset] = static_cast<uint8_t>(source.currenct_format);
+
+	if (!found) {
+		// Fallback - use the currency code instead
+		memcpy(&destination[InfoOffsetCurrencySymbol],
+		       source.currency_code.c_str(),
+		       source.currency_code.length());
+
+		// Force separation between symbol and amount
+		bit::set(destination[offset], bit::literals::b0);
+	}
+
+	offset = InfoOffsetCurrencyPrecision;
+	destination[offset] = source.currency_precision;
+
+	// Mark locale as generated
+
+	config.is_locale_generated = true;
+
+	// If locale changed, log it
+
+	static bool already_logged = false;
+
+	static Country logged_country            = {};
+	static LocalePeriod logged_real_period   = {};
+	static LocalePeriod logged_config_period = {};
+
+	if ((logged_country == config.country) && already_logged &&
+	    (logged_real_period == locale_period) &&
+	    (logged_config_period == config.locale_period)) {
+		return;
+	}
+
+	auto period_to_string = [](const LocalePeriod period) {
+		if (period == LocalePeriod::Modern) {
+			return "modern";
+		} else {
+			return "historic";
+		}
+	};
+
+	std::string additional_comment = {};
+	if (locale_period != config.locale_period) {
+		additional_comment = " (";
+		additional_comment += period_to_string(config.locale_period);
+		additional_comment += " locale not known)";
+	}
+
+	LOG_MSG("DOS: Loaded %s locale for country %d, '%s'%s",
+	        period_to_string(locale_period),
+	        static_cast<int>(config.country),
+	        country_info.country_name.c_str(),
+	        additional_comment.c_str());
+
+	already_logged       = true;
+	logged_country       = config.country;
+	logged_real_period   = locale_period;
+	logged_config_period = config.locale_period;
```

Comment:
This is a huge function. Can't it be broken up into smaller logical chunks and/or lambdas with descriptive names and well-defined small scopes?

Reply:
OK, split it into more functions.